tuner/libtuner.py

```diff
@@ -0,0 +1,1302 @@
+    )
+    handle_error(
+        condition=(compiling_rate < 10),
+        msg=f"Compiling rate [{compiling_rate:.1f}%] < 10%",
+        level=logging.WARNING,
+    )
+
+    collision_detected, unique_indexes = collision_handler(
+        compiled_candidates_hash_list
+    )
+    if collision_detected:
+        logging.critical(f"Remains [{len(unique_indexes)}] unique candidate indexes")
+
+    return compiled_candidates if not collision_detected else unique_indexes
+
+
+def parse_dispatch_benchmark_results(
+    path_config: PathConfig,
+    benchmark_results: list[TaskResult],
+    candidate_trackers: list[CandidateTracker],
+) -> tuple[list[ParsedDisptachBenchmarkResult], list[str]]:
+    benchmark_result_configs = []
+    dump_list = []
+    incomplete_list = []
+
+    for benchmark_result in benchmark_results:
+        res_str = benchmark_result.result.stdout
+        candidate_id = benchmark_result.candidate_id
+        res = IREEBenchmarkResult(candidate_id, res_str)
+        benchmark_time = res.get_mean_time()
+        if benchmark_time is None:
+            incomplete_list.append(candidate_id)
+            continue
+        assert benchmark_time is not None
+        candidate_trackers[candidate_id].first_benchmark_time = benchmark_time
+        candidate_trackers[
+            candidate_id
+        ].spec_path = path_config.specs_dir / path_config.get_candidate_spec_filename(
+            candidate_id
+        )
+        mlir_path = candidate_trackers[candidate_id].dispatch_mlir_path
+        spec_path = candidate_trackers[candidate_id].spec_path
+        assert mlir_path is not None and spec_path is not None
+        dump_list.append(generate_display_DBR(candidate_id, benchmark_time) + "\n")
+
+        benchmark_result_configs.append(
+            (
+                ParsedDisptachBenchmarkResult(
+                    candidate_id,
+                    benchmark_time,
+                    mlir_path,
+                    spec_path,
+                )
+            )
+        )
+
+    if incomplete_list:
+        dump_list += [f"Candidate {i} not incompleted" for i in incomplete_list]
+
+    return benchmark_result_configs, dump_list
+
+
+def generate_sample_task_result(
+    stdout: str, candidate_id: int, device_id: str
+) -> TaskResult:
+    res = subprocess.CompletedProcess(
+        args=[""],
+        stdout=stdout,
+        returncode=0,
+    )
+    return TaskResult(result=res, candidate_id=candidate_id, device_id=device_id)
+
+
+def generate_dryrun_dispatch_benchmark_results(
+    compiled_candidates: list[int],
+) -> list[TaskResult]:
+    logging.info("generate_dryrun_dispatch_benchmark_results")
+
+    task_results = [
+        generate_sample_task_result(
+            f"process_time/real_time_mean    {random.uniform(100.0, 500.0):.3g} ms",
+            i,
+            str(0),
+        )
+        for i in compiled_candidates
+    ]
+
+    return task_results
```

Comment:
More wondering about use of `random` here. I'd rather show some obvious placeholder than show truly random data.

Reply:
Can you show me some examples how you'd want to change?

Reply:
In a dry run, you're not actually running benchmarks, so why generate any benchmark results at all? If you do for some reason need to generate results, I'd make it obvious that the results are synthetic, like `process_time/real_time_mean    DRY_RUN`.

Reply:
The existing benchmarking functions still rely on using regex parsing method, extracts numbers from result_str, and does some order sorting and reconstruct output string for display purpose. The strings generated in this function are not the final form, these are fake results that pretend to be the captured console text, and these results will be processed again later in other functions (e.g. result normalization, calibrate the data and calculate the perf improvement percent)

Reply:
Ah, so more issues stemming from passing strings around instead of structured data. A tale as old as time 😅 

Okay for now but it would help to refactor such that fake results are not needed.